src/storage/FileResourceStore.ts

```diff
@@ -0,0 +1,427 @@
+
+  private readonly predicates = {
+    aType: DataFactory.namedNode(`${RDF}type`),
+    modified: DataFactory.namedNode(`${TERMS}modified`),
+    contains: DataFactory.namedNode(`${LDP}contains`),
+    mtime: DataFactory.namedNode(`${STAT}mtime`),
+    size: DataFactory.namedNode(`${STAT}size`),
+  };
+
+  private readonly objects = {
+    container: DataFactory.namedNode(`${LDP}Container`),
+    basicContainer: DataFactory.namedNode(`${LDP}BasicContainer`),
+    ldpResource: DataFactory.namedNode(`${LDP}Resource`),
+    dateTime: DataFactory.namedNode(`${XML}dateTime`),
+  };
+
+  /**
+   * @param baseRequestURI - Will be stripped of all incoming URIs and added to all outgoing ones to find the relative
+   * path.
+   * @param rootFilepath - Root filepath in which the resources and containers will be saved as files and directories.
+   * @param interactionController - Instance of InteractionController to use.
+   */
+  public constructor(baseRequestURI: string, rootFilepath: string, interactionController: InteractionController) {
+    this.baseRequestURI = trimTrailingSlashes(baseRequestURI);
+    this.rootFilepath = trimTrailingSlashes(rootFilepath);
+    this.interactionController = interactionController;
+  }
+
+  /**
+   * Store the incoming data as a file under a file path corresponding to `container.path`,
+   * where slashes correspond to subdirectories.
+   * @param container - The identifier to store the new data under.
+   * @param representation - Data to store. Only File streams are supported.
+   *
+   * @returns The newly generated identifier.
+   */
+  public async addResource(container: ResourceIdentifier, representation: Representation): Promise<ResourceIdentifier> {
+    if (representation.dataType !== DATA_TYPE_BINARY) {
+      throw new UnsupportedMediaTypeHttpError('FileResourceStore only supports binary representations.');
+    }
+    const path = this.parseIdentifier(container);
+    const { slug, raw } = representation.metadata;
+    const linkTypes = representation.metadata.linkRel?.type;
+    let metadata;
+    if (raw.length > 0) {
+      metadata = streamifyArray(raw).pipe(new StreamWriter({ format: 'text/turtle' }));
+    }
+
+    const isContainer = this.interactionController.isContainer(slug, linkTypes);
+    const newIdentifier = this.interactionController.generateIdentifier(isContainer, slug);
+    if (!isContainer) {
+      // Create a file for the resource with as filepath the parent container.
+      return this.createFile(path, newIdentifier, representation.data, path.endsWith('/'), metadata);
+    }
+
+    // Create a new container as subdirectory of the parent container.
+    return this.createContainer(path, newIdentifier, path.endsWith('/'), metadata);
+  }
+
+  /**
+   * Deletes the given resource.
+   * @param identifier - Identifier of resource to delete.
+   */
+  public async deleteResource(identifier: ResourceIdentifier): Promise<void> {
```

Comment:
I would also suggest to break up this function into multiple functions, cfr. my comment on `addResource`. Then you can see the high-level structure (basically 3 or so functions being called) with it's if/else logic, and the details per function.